compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -459,13 +457,184 @@ struct DistributeBroadcast final : OpDistributionPattern<vector::BroadcastOp> {
   }
 };
 
+static int64_t getShuffleOffset(NestedLayoutAttr layout, int64_t dim) {
+  // Get strides for dimensions based on layouts.
+  SmallVector<int64_t> threadBasis(layout.getThreadBasis());
+  SmallVector<int64_t> basisStrides(threadBasis.size(), 1);
+  // Take prefix sum to get strides.
+  for (int i = basisStrides.size() - 1; i >= 1; --i) {
+    basisStrides[i - 1] *= threadBasis[i];
+  }
+  // Remove non-active thread ids.
+  SmallVector<int64_t> activeThreadStrides;
+  for (auto [i, stride] : llvm::enumerate(basisStrides)) {
+    if (layout.getThreadActiveIds()[i]) {
+      activeThreadStrides.push_back(stride);
+    }
+  }
+  // Apply permutation on the active thread strides.
+  // TODO: Do we need to do inversion or not?
+  applyPermutationToVector(activeThreadStrides, layout.getThreadOrder());
+  return activeThreadStrides[dim];
+}
+
+static int64_t getShuffleWidth(NestedLayoutAttr layout, int64_t dim) {
+  return layout.getThreadsPerOuter()[dim];
+}
+
+struct DistributeMultiReduction final
+    : OpDistributionPattern<vector::MultiDimReductionOp> {
+  using OpDistributionPattern::OpDistributionPattern;
+
+  DistributeMultiReduction(MLIRContext *context, int64_t maxBitsPerShuffle,
+                           int64_t benefit = 1)
+      : OpDistributionPattern(context, benefit),
+        maxBitsPerShuffle(maxBitsPerShuffle) {}
+
+  LogicalResult matchAndRewrite(vector::MultiDimReductionOp multiReduceOp,
+                                DistributionSignature &signature,
+                                PatternRewriter &rewriter) const override {
+    VectorValue srcVector = multiReduceOp.getSource();
+    VectorValue accVector = dyn_cast<VectorValue>(multiReduceOp.getAcc());
```

Comment:
ubernit
```suggestion
    auto accVector = dyn_cast<VectorValue>(multiReduceOp.getAcc());
```

also below